txnbuild/transaction.go

```
@@ -406,3 +406,89 @@ func (tx *Transaction) SignWithKeyString(keys ...string) error {
 
 	return tx.Sign(signers...)
 }
+
+// VerifyChallengeTx is a factory method that verifies a SEP 10 challenge transaction,
+// for use in web authentication.
+// More details on SEP 10: https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0010.md
+func VerifyChallengeTx(challengeTx, serverAccountId, network string) (bool, error) {
+	tx, err := TransactionFromXDR(challengeTx)
+	if err != nil {
+		return false, err
+	}
+	tx.Network = network
+
+	// verify transaction source
+	if tx.SourceAccount != nil {
+		if tx.SourceAccount.GetAccountID() != serverAccountId {
+			return false, errors.New("transaction source account is not equal to server's account")
+		}
+	} else {
+		return false, errors.New("transaction requires a source account")
+	}
+
+	// verify timebounds
+	if tx.Timebounds.wasBuilt {
+		// if timebound is not infinite
+		if tx.Timebounds.MinTime > 0 && tx.Timebounds.MaxTime > 0 {
+			currentTime := time.Now().UTC().Unix()
+			if currentTime < tx.Timebounds.MinTime || currentTime > tx.Timebounds.MaxTime {
+				return false, errors.New("transaction timebounds has elapsed")
+			}
+		}
+	} else {
+		return false, errors.New("transaction requires timebounds")
+	}
+
+	// verify operation
+	if len(tx.Operations) == 1 {
```

Again, reverse. Fail fast if `len(tx.Operations) != 1`